SciLean/Data/Mesh/SurfaceMesh.lean

```diff
@@ -0,0 +1,350 @@
+/-
+Copyright (c) 2023 Siddharth Bhat. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddharth Bhat
+
+Surface Mesh based on discrete differential geometry.
+https://github.com/GeometryCollective/geometry-processing-js/blob/c7ea47ae808979a5361e3fcd61bf921a194bf897/core/mesh.js#L3-L213
+-/
+
+import SciLean.Core
+
+
+-- abbrev V3 : Type := ℝ^{3}
+-- 3D vector.
+structure V3 where
+  x : ℝ
+  y : ℝ
+  z : ℝ
+
+/- index of vertices --/
+abbrev Index (name : Name) := Nat
+abbrev Index.invalid : Index α := 424242
+
+structure Vertex where
+  index : Index `Vertex := .invalid
+  halfedge : Index `Halfedge := .invalid
+deriving Inhabited
+
+/-- Half edge data structure -/
+structure Halfedge where
+  index : Index `Halfedge := .invalid
+  edge : Index `Edge := .invalid
+  twin : Option (Index `Halfedge) := .none
+  vertex : Index `Vertex := .invalid
+  face : Index `Face := .invalid
+  corner : Option (Index `Corner) := .none
+  onBoundary : Bool := false
+deriving Inhabited
+
+structure Face where
+  index : Index `Face := .invalid
+  halfedge : Index `Halfedge := .invalid
+deriving Inhabited
+
+structure Corner where
+  index : Index `Corner := .invalid
+  halfedge : Index `Halfedge := .invalid
+deriving Inhabited
+
+structure Edge where
+  index : Index `Edge := .invalid
+  halfedge : Index `Halfedge := .invalid
+deriving Inhabited
+
+structure Boundary where
+  gens : Array (Index `Halfedge) := #[]
+deriving Inhabited
+
+/-
+ * @property {module:Core.Vertex[]} vertices The vertices contained in this mesh.
+ * @property {module:Core.Edge[]} edges The edges contained in this mesh.
+ * @property {module:Core.Face[]} faces The faces contained in this mesh.
+ * @property {module:Core.Corner[]} corners The corners contained in this mesh.
+ * @property {module:Core.Halfedge[]} halfedges The halfedges contained in this mesh.
+ * @property {module:Core.Face[]} boundaries The boundary loops contained in this mesh.
+ * @property {Array.<module:Core.Halfedge[]>} generators An array of halfedge arrays, i.e.,
+ -/
+structure SurfaceMesh where
+  positions  : Array V3 := {}
+  vertices   : Array Vertex := {}
+  edges      : Array Edge := {}
+  faces      : Array Face := {}
+  corners    : Array Corner := {}
+  halfedges  : Array Halfedge := {}
+  boundaries : Array Boundary := {}
+  -- generators : Array () [TODO: implement]
```

Comment:
How is it with Arrays here? Are all the values in them boxed? That effectively leads to double memory footprint and the actual values can be scattered all around the memory. Or is lean doing something smart for small structures? 

I would suggest using my DataArray which dumps everything into one ByteArray but I do not have working Deriving for PlainDataType class, so it would be a pain to use for you.

Reply:
That is a good point. I am afraid that indeed, yes, these will all be boxed. Shall we making switching to `DataArray` a separate PR, where we can also correctly implement `Deriving`?

Reply:
That is a good point. I am afraid that indeed, yes, these will all be boxed. Shall we making switching to `DataArray` a separate PR, where we can also correctly implement `Deriving`?

Reply:
Probably not high priority right now, but might be worth doing sooner rather than later.

Also I would maybe rename PlainDataType to Haskell's Storable as it is similar.